robot_modeling/DQ_SerialManipulator.m

```diff
@@ -152,188 +232,34 @@ function set_effector(obj,effector)
-                h(5) = -d2*h4-a2*h2;
-                h(6) = -d2*h3+a2*h1;
-                h(7) = -(d2*h2+a2*h4);
-                h(8) = d2*h1-a2*h3;
-            end
-            
-            dq = DQ(h);
-        end
-        
-        function p = get_z(~,h)
-            p(1) = 0;
-            p(2) = h(2)*h(4) + h(1)*h(3);
-            p(3) = h(3)*h(4) - h(1)* h(2);
-            p(4) = (h(4)^2-h(3)^2-h(2)^2+h(1)^2)/2;
-            p(5) = 0;
-            p(6) = h(2)*h(8)+h(6)*h(4)+h(1)*h(7)+h(5)*h(3);
-            p(7) = h(3)*h(8)+h(7)*h(4)-h(1)*h(6)-h(5)*h(2);
-            p(8) = h(4)*h(8)-h(3)*h(7)-h(2)*h(6)+h(1)*h(5);
-        end
-        
-        function J = raw_pose_jacobian(obj,q,ith)
-            % RAW_POSE_JACOBIAN(q) returns the Jacobian that satisfies 
-            % vec(x_dot) = J * q_dot, where x = fkm(q) and q is the 
-            % vector of joint variables.
-            %
-            % RAW_POSE_JACOBIAN(q,ith) returns the Jacobian that
-            % satisfies vec(x_ith_dot) = J * q_dot(1:ith), where 
-            % x_ith = fkm(q, ith), that is, the fkm up to the i-th link.
-            %
-            % This function does not take into account any base or
-            % end-effector displacements and should be used mostly
-            % internally in DQ_kinematics
-            
-            if nargin == 3
-                n = ith;
-                x_effector = obj.raw_fkm(q,ith);
+                x = obj.reference_frame_*obj.raw_fkm(q, ith); %Takes into account the base displacement
             else
-                n = obj.n_links;
-                x_effector = obj.raw_fkm(q);
+                x = obj.reference_frame_*obj.raw_fkm(q)*obj.curr_effector_;
             end
-            
-            x = DQ(1);
-            J = zeros(8,n);
-            
-            for i = 0:n-1
-                % Use the standard DH convention
-                if strcmp(obj.convention,'standard')
-                    z = DQ(obj.get_z(x.q));
-                else % Use the modified DH convention
-                    w = DQ([0,0,-sin(obj.alpha(i+1)),cos(obj.alpha(i+1)),0, ...
-                        0,-obj.a(i+1)*cos(obj.alpha(i+1)), ...
-                        -obj.a(i+1)*sin(obj.alpha(i+1))] );
-                    z = 0.5*x*w*x';
-                end
-                
-                x = x*obj.dh2dq(q(i+1),i+1);
-                j = z * x_effector;
-                J(:,i+1) = vec8(j);
-            end
-        end
-        
+        end      
+             
         function J = pose_jacobian(obj, q, ith)
             % POSE_JACOBIAN(q) returns the Jacobian that satisfies
             % vec(x_dot) = J * q_dot, where x = fkm(q) and
             % q is the vector of joint variables. It takes into account
             % both base and end-effector displacements (their default
             % values are 1).
             
-            if nargin == 3 && ith < obj.n_links
+            if nargin == 3 && ith < obj.dim_configuration_space_
```

Comment:
Again, I implemented it as in the C++ version. But, If I am not wrong about @mmmarinho's logic, in that line of the code, we want to know if the index (valid or not) is lower than the robot configuration space. We are checking if both the joint positions and the index are valid in the concrete classes (DQ_SerialManipulatorDH and DQ_SerialManipulatorMDH).
For instance, if we have: 
```Matlab
robot = KukaLwr4Robot.kinematics;
q = [0 0 0 0 0 0 0];
J = robot.pose_jacobian(q, 7);
```
The method _obj.pose_jacobian(obj, q, 7)_ will run the second condition (the code below the statement _else_) despite the method having 3 arguments and using a valid index.

@mmmarinho if I am wrong, please correct me.

Reply:
@juanjqo 
That specific line of code is unique to MATLAB... So not much to do with the CPP implementation IMO.

`ith` will not always exist in the context of the method, because it will depend on which signature you call.
`pose_jacobian(q)` will have `nargin<3`
`pose_jacobian(q,7)` for instance will have `nargin==3`

You can do as @bvadorno  suggested, and do something like

```matlab
if nargin == 3
 _check_ith_link(ith)
%do whatever
endif
```